File: modules/examples/cyclonedds/cpp/publisher.cpp
#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <string>
#include <thread>

#include <dds/dds.hpp>

#include "HelloMessage.hpp"

namespace {

constexpr auto kTopicName = "robotics_lab.examples.hello";

std::uint32_t message_count(int argc, char* argv[]) {
  if (argc < 2) {
    return 5;
  }
  return static_cast<std::uint32_t>(std::stoul(argv[1]));
}

}  // namespace

int main(int argc, char* argv[]) {
  try {
    const auto count = message_count(argc, argv);
    dds::domain::DomainParticipant participant(
        org::eclipse::cyclonedds::domain::default_id());
    dds::topic::Topic<robotics_lab_examples::HelloMessage> topic(
        participant, kTopicName);
    dds::pub::Publisher publisher(participant);
    dds::pub::DataWriter<robotics_lab_examples::HelloMessage> writer(publisher,
                                                                     topic);

    std::this_thread::sleep_for(std::chrono::seconds(1));
    for (std::uint32_t sequence = 1; sequence <= count; ++sequence) {
      robotics_lab_examples::HelloMessage message(
          "cpp-publisher", sequence, "Hello from Cyclone DDS C++");
      writer.write(message);
      std::cout << "published sequence=" << sequence << " sender="
                << message.sender() << '\n';
      std::this_thread::sleep_for(std::chrono::milliseconds(200));
    }
  } catch (const std::exception& error) {
    std::cerr << "publisher failed: " << error.what() << '\n';
    return EXIT_FAILURE;
  }

  return EXIT_SUCCESS;
}

File: modules/examples/cyclonedds/python/publisher.py
#!/usr/bin/env python3
"""Publish example messages with Cyclone DDS Python."""

from __future__ import annotations

import argparse
import time
from collections.abc import Sequence

from cyclonedds.domain import DomainParticipant
from cyclonedds.pub import DataWriter
from cyclonedds.topic import Topic
from robotics_lab_examples import HelloMessage


TOPIC_NAME = "robotics_lab.examples.hello"


def parse_args(argv: Sequence[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("count", nargs="?", type=int, default=5)
    return parser.parse_args(argv)


def main(argv: Sequence[str] | None = None) -> int:
    args = parse_args(argv)
    participant = DomainParticipant()
    topic = Topic(participant, TOPIC_NAME, HelloMessage)
    writer = DataWriter(participant, topic)

    time.sleep(1)
    for sequence_number in range(1, args.count + 1):
        message = HelloMessage(
            sender="python-publisher",
            sequence_number=sequence_number,
            text="Hello from Cyclone DDS Python",
        )
        writer.write(message)
        print(
            f"published sequence={message.sequence_number} "
            f"sender={message.sender}",
            flush=True,
        )
        time.sleep(0.2)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
